src/server/pps/server/api_server.go

```diff
@@ -1598,16 +1504,36 @@ func (a *apiServer) writePipelineInfoToFileset(ctx context.Context, pipelineInfo
 	return resp.FilesetId, nil
 }
 
+// commitPipelineInfoFromFileset is a helper for all pipeline updates that
+// creates a commit with 'pipelineInfo' in SpecRepo (in PFS). It's called in
+// both the case where a user is updating a pipeline and the case where a user
+// is creating a new pipeline.
+// pipelineInfo.Version is used to ensure that (in the absence of
+// transactionality) the new pipelineInfo is only applied on top of the previous
+// pipelineInfo - if the version of the pipeline has changed, this operation
+// will error out.
 func (a *apiServer) commitPipelineInfoFromFileset(
 	txnCtx *txncontext.TransactionContext,
 	pipelineName string,
 	filesetID string,
-	prevSpecCommit *pfs.Commit,
+	prevPipelineVersion uint64,
 ) (*pfs.Commit, error) {
+	// Make sure the StoredPipelineInfo is the same version we expect so we don't clobber another pipeline update
+	if prevPipelineVersion != 0 {
+		latestPipelineInfo, err := a.latestPipelineInfo(txnCtx, pipelineName)
```

Comment:
We should restructure this such that the not found error propagates to here, rather than returning nil, nil in that case.